hail/src/main/scala/is/hail/expr/ir/EmitCodeBuilder.scala

```diff
@@ -128,18 +128,31 @@ class EmitCodeBuilder(val emb: EmitMethodBuilder[_], var code: Code[Unit]) exten
               s"\n  expected ${ pcpt.pt }")
           pc.codeTuple()
         case (EmitParam(ec), EmitParamType(pt)) =>
-          if (ec.pt != pt)
-            throw new RuntimeException(s"invoke ${ callee.mb.methodName }: arg $i: type mismatch:" +
-              s"\n  got ${ ec.pt }" +
-              s"\n  expected ${ pt }")
-          if (ec.pt.required) {
-            append(ec.setup)
-            append(Code.toUnit(ec.m))
-            ec.codeTuple()
+          if (!ec.pt.equalModuloRequired(pt)) {
+            throw new RuntimeException(s"invoke ${callee.mb.methodName}: arg $i: type mismatch:" +
+              s"\n  got ${ec.pt}" +
+              s"\n  expected ${pt}")
+          }
+
+          val castEc = (ec.pt.required, pt.required) match {
+            case (true, false) => {
+              ec.map(pc => PCode(pc.pt.setRequired(pt.required), pc.code))
+            }
+            case (false, true) => {
+              EmitCode.fromI(callee) { cb => IEmitCode.present(this, ec.toI(this).get(this))}
+            }
+            case _ => ec
```

Comment:
get rid of this base case, it's dead code

Reply:
er, nvm

Reply:
I wrote the cast out separately like this, figured it'd be easier to delete this cast block and go back to just using `ec` when we fix the underlying problem.